libraries/aws/common/test/unit/aws_iot_tests_doc_parser.c

```diff
@@ -0,0 +1,280 @@
+{
+    /* Parse JSON document with string, int, bool, and null. */
+    {
+        const char pJsonDocument[ 82 ] = "{\"name\" \n\r:\n \"John Smith\", \"age\"    :\n\r 30, \n \"isAlive\"  : true, \r \"spouse\":null}";
+        size_t jsonDocumentLength = strlen( pJsonDocument );
+
+        _parseJson( true, pJsonDocument, jsonDocumentLength, "name", "\"John Smith\"", 12 );
+        _parseJson( true, pJsonDocument, jsonDocumentLength, "age", "30", 2 );
+        _parseJson( true, pJsonDocument, jsonDocumentLength, "isAlive", "true", 4 );
+        _parseJson( true, pJsonDocument, jsonDocumentLength, "spouse", "null", 4 );
+
+        /* Attempt to find a key not in the JSON document. */
+        _parseJson( false, pJsonDocument, jsonDocumentLength, "address", NULL, 0 );
+    }
+
+    /* Parse JSON document with objects and arrays. */
+    {
+        const char pJsonDocument[ 91 ] = "{\"object\" : { \"nestedObject\": { \"string\":\"\\\"test\\\"\", \"array\":[[1,2,3],[1,2,3],[1,2,3]]}}}";
+        size_t jsonDocumentLength = strlen( pJsonDocument );
+
+        _parseJson( true, pJsonDocument, jsonDocumentLength, "object", "{ \"nestedObject\": { \"string\":\"\\\"test\\\"\", \"array\":[[1,2,3],[1,2,3],[1,2,3]]}}", 76 );
+        _parseJson( true, pJsonDocument, jsonDocumentLength, "nestedObject", "{ \"string\":\"\\\"test\\\"\", \"array\":[[1,2,3],[1,2,3],[1,2,3]]}", 57 );
+        _parseJson( true, pJsonDocument, jsonDocumentLength, "array", "[[1,2,3],[1,2,3],[1,2,3]]", 25 );
+    }
+
+    /* JSON document with escape sequences. */
+    {
+        const char pJsonDocument[ 40 ] = "{\"key\": \"value\", \"ke\\\"y2\": \"\\\"value\\\"\"}";
+        size_t jsonDocumentLength = strlen( pJsonDocument );
+
+        /* Attempt to find a JSON key that is actually a value. */
+        _parseJson( false, pJsonDocument, jsonDocumentLength, "value", NULL, 0 );
+
+        /* Attempt to find a JSON key that is a substring of a key. */
+        _parseJson( false, pJsonDocument, jsonDocumentLength, "ke", NULL, 0 );
+
+        /* Find a key and string that contain escaped quotes. */
+        _parseJson( true, pJsonDocument, jsonDocumentLength, "ke\\\"y2", "\"\\\"value\\\"\"", 11 );
+    }
+
+    /* Short JSON document. */
+    {
+        const char pJsonDocument[ 16 ] = "{\"key\":\"value\"}";
+        size_t jsonDocumentLength = strlen( pJsonDocument );
+
+        /* Attempt to find a key longer than the JSON document. */
+        _parseJson( false, pJsonDocument, jsonDocumentLength, "longlonglonglongkey", NULL, 0 );
+    }
+
+    /* JSON with '{' '}' in value in nested level*/
+    {
+        const char pJsonDocument[ 42 ] = "{\"key\":{\"key2\":\"{{{{}}\", \"key3\":\"value\"}}";
+        size_t jsonDocumentLength = strlen( pJsonDocument );
+
+        /* Attempt to find a key longer than the JSON document. */
+        _parseJson( true, pJsonDocument, jsonDocumentLength, "key", "{\"key2\":\"{{{{}}\", \"key3\":\"value\"}", 33 );
+    }
```

Comment:
Other suggested test cases:

1. Object as element of array
1. KVP with value `false`
1. Arrays with multiple different element types, e.g.

```
[ 4, "a-string", { "foo" : "bar", "arr" : [ 1, 2, 7] }, 99 ]
```